### ISSUE TITLE
Bug: [no-unnecessary-type-assertion] false positive on `reduce` initial value behind optional chaining

### ISSUE THREAD

**Issue by kyriiherman** [oxc-project/tsgolint#1053]
### Summary

`typescript/no-unnecessary-type-assertion` (type-aware) reports a **load-bearing** type assertion as unnecessary when a `reduce` initial-value cast is called behind an **optional chain**. Removing the flagged assertion changes the program's types and breaks type-checking (`tsc` reports `TS7053`), so the assertion is not unnecessary.

This mirrors an upstream typescript-eslint false positive — see typescript-eslint/typescript-eslint#12485. Since tsgolint syncs this rule's behavior/tests with typescript-eslint, filing here to track porting the fix once upstream resolves.

### Reproduction

`repro.ts`:

```ts
declare const items: string[] | undefined;

// Flagged as unnecessary, but the cast is load-bearing: it seeds the
// accumulator generic of `reduce`. Removing it makes `acc` infer as `{}`.
const counts = items?.reduce((acc, item) => {
  acc[item] = (acc[item] ?? 0) + 1;
  return acc;
}, {} as Record<string, number>);

// Contrast: the SAME reduce without optional chaining is NOT flagged.
declare const items2: string[];
const counts2 = items2.reduce((acc, item) => {
  acc[item] = (acc[item] ?? 0) + 1;
  return acc;
}, {} as Record<string, number>);
```

`.oxlintrc.json`:

```json
{
  "plugins": ["typescript"],
  "categories": { "correctness": "off" },
  "rules": { "typescript/no-unnecessary-type-assertion": "error" }
}
```

`tsconfig.json`:

```json
{ "compilerOptions": { "strict": true } }
```

Run:

```sh
oxlint --type-aware repro.ts
```

### Actual

```
repro.ts:8:3: error typescript(no-unnecessary-type-assertion): This assertion is unnecessary since the receiver accepts the original type of the expression.
```

Only the optional-chained `reduce` (line 8) is flagged; the identical non-nullable `items2.reduce(...)` is correctly left alone.

### Expected

No diagnostic. `{} as Record<string, number>` seeds the generic accumulator type parameter `U` of `Array.prototype.reduce<U>(cb, initialValue: U)`. Without the assertion, `U` infers as `{}` and the callback no longer compiles:

```
error TS7053: Element implicitly has an 'any' type because expression of type
'string' can't be used to index type '{}'.
```

Because removing the assertion changes the program's types, it is load-bearing and should not be reported.

### Root cause

The "receiver accepts the original type of the expression" check treats `{}` as assignable to the accumulator type `Record<string, number>` — which is true — but that receiver type is the type parameter as already inferred *with* the assertion present. The check doesn't account for the assertion being what pins that inference, so it is reported as redundant. This is the same "unnecessary from assignment, but required for inference" class as typescript-eslint/typescript-eslint#12250; the optional-chaining variant is still open upstream (#12485).

### Environment

- oxlint: 1.71.0
- oxlint-tsgolint: 0.24.0
- Also reproduces in @typescript-eslint/eslint-plugin 8.62.1 (TypeScript 5.9.3 and 6.0.3) — see #12485.

**Comment by camc314** [comment]
fixed by https://github.com/oxc-project/tsgolint/pull/1183
